Warp10

aurrelhebert opened the pull request:
### Required for all PRs:
- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] README.md updated (if adding a new plugin)

BlackYoup commented:
Hi,

Any news on this ? I would love to see this merged :)

danielnelson commented:
Right now this is pretty low priority, best way to get it noticed is by adding a :+1: to the top comment or by [contacting sales](https://www.influxdata.com/contact-sales/)

aurrelhebert commented:
Hi,

Any news on this ?

whitetigle commented:
👍 
Any news?

waxzce commented:
If it can help: as Clever Cloud is using telegraf and warp10 together on large production scale, we will maintain this integration.

aurrelhebert commented:
Hello, if it can help too, us at OVH are also using telegraf on large production scale and will contribute in any mean to maintain it (review, update, ...)

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,179 @@
+	req.Header.Set("X-Warp10-Token", o.Token)
+	req.Header.Set("Content-Type", "text/plain")
+
+	client := &http.Client{}
```
We need to create a single client+transport instead of using a new one or the default each Write.  Check out the http output for an example.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,179 @@
+	index := 0
+	for _, mm := range metrics {
+
+		for k, v := range mm.Fields() {
```
Let's switch over to FieldList her for performance reasons.

danielnelson commented on plugins/outputs/warp10/warp10_test.go:
```diff
@@ -0,0 +1,22 @@
+package warp10
```
We need to add some more substantial tests, particularly around the creation of the request body.

aurrelhebert commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,179 @@
+	case int64:
+		retv = IntToString(int64(p))
+	case string:
+		retv = fmt.Sprintf("'%s'", p)
```
The `'` character is required in Warp10 ingest format to push any string data

danielnelson replied:
What about in the string though?  For example what if `p := "bob's"`?

I found this example `// dd{} 'it%27s ok'`, if the only char that needs replaced is the quote then you could use a strings.Replacer.

aurrelhebert commented:
Thanks a lot for you review! I just pushed an update, let me know about it :). 
I updated to use it more nicely the HTTP server and to handle errors. 
I removed previous UnitTest and test only the body payload.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,204 @@
+
+			metricValue, err := buildValue(field.Value)
+			if err != nil {
+				log.Printf("Warp: %s\n", err.Error())
```
Use this style:
```
log.Printf("E! [outputs.warp10] Could not encode value: %v", err)
```

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,204 @@
+			}
+			metric.Value = metricValue
+
+			tagsSlice := buildTags(mm.Tags())
```
Sorry I missed this before, we will want to use TagList() here for the same reasons we switched to FieldList(), its quite a bit more performant (and we will be removing this method soon).

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,204 @@
+// GenWarp10Payload compute Warp 10 metrics payload
+func (w *Warp10) GenWarp10Payload(metrics []telegraf.Metric, now time.Time) string {
+	collectString := make([]string, 0)
+	index := 0
```
I don't think this is used.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	if w.Timeout.Duration == 0 {
+		w.Timeout.Duration = defaultClientTimeout
```
Move above createClient, or maybe just into the createClient function.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,204 @@
+		retv = floatToString(float64(p))
+	default:
+		retv = fmt.Sprintf("'%s'", p)
+		//return "", fmt.Errorf("unexpected type %T with value %v for Warp", v, v)
```
Remove any commented out code.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,204 @@
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 300 {
+		body, _ := ioutil.ReadAll(resp.Body)
```
I'd like to improve this error a bit, do you have an example of the response body on an error?

aurrelhebert replied:
**Case of an invalid token**:
> 2019-08-24T09:53:16Z E! [agent] Error writing to output [warp10]: http://127.0.0.1:8090: <html>
> <head>
> <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
> <title>Error 500 io.warp10.script.WarpScriptException: Invalid token.</title>
> </head>
> <body><h2>HTTP ERROR 500</h2>
> <p>Problem accessing /api/v0/update. Reason:
> <pre>    io.warp10.script.WarpScriptException: Invalid token.</pre></p>
> </body>
> </html>

**Case of an unvalid input**:
> 2019-08-24T09:53:16Z E! [agent] Error writing to output [warp10]: http://127.0.0.1:8090: <html>
> <head>
> <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
> <title>Error 500 Parse error at &apos;1566640396329990// telegraf{} 5 test.processes.idle{host=laptop1134849.home,source=telegraf} 0&apos;</title>
> </head>
> <body><h2>HTTP ERROR 500</h2>
> <p>Problem accessing /api/v0/update. Reason:
> <pre>    Parse error at &apos;1566640396329990// telegraf{} 5 test.processes.idle{host=laptop1134849.home,source=telegraf} 0&apos;</pre></p>
> </body>
> </html>

aurrelhebert replied:
I just added all error message I know about that can be returned by Warp10. As well as some test body examples. 
Warp is sending only 200 as return code in case of success, I replaced the test to use an equality check here.

danielnelson replied:
I'm not entirely satisfied with the error handling yet.  Using the docker container I poked around a bit at the API, and I noticed that the error string is always included in the response status line, here are few examples:
```
HTTP/1.1 500 Error when updating data: Invalid token.
HTTP/1.1 500 Error when updating data: Parse error at '/ test{} 42' (Missing location/elevation separator.) (Missing location/elevation separator.)
```

I also noticed that the status line does not use HTML escaping, which is a bit nicer especially in the case of an unexpected error.  I think we should use the status line as the source for the error strings.  We can split it into three pieces using strings.SplitN, the third part will be the error string.

danielnelson commented on plugins/outputs/warp10/README.md:
````diff
@@ -0,0 +1,19 @@
+
+```
+[[outputs.warp10]]
+warpUrl = "http://localhost:4242/api/v0/update"
````
I think we should have the user only specify the base URL `http://localhost:4242`, then we can tack on `/api/v0/update`.  This will be a bit easier for the user and matches other plugins.  However, I wouldn't do this if its common for the path to differ, for example due to proxy use.

What about TLS configuration?  We generally have this for all HTTP plugins and we try to support it in all new plugins, does Warp10 support this?

aurrelhebert commented:
Thanks again for your review, let me know if you want to update the error messages. Using the body returned they stay similar to the one returned by a Warp10 platform.

danielnelson commented:
Thank @aurrelhebert, I just replied above regarding the [errors](https://github.com/influxdata/telegraf/pull/1923#discussion_r318847901).  I think we need to do this along with adding a few more unit tests, this is especially important since I don't have a Warp10 account.  If you are having trouble getting started on the tests I suggest checking out the `http` output for some similar examples.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,293 @@
+
+var sampleConfig = `
+  # prefix for metrics class Name
+  prefix = "Prefix"
```
Based on the style of some built in metrics, and because you probably want to end this with a `.`, let's set this to `prefix = "telegraf."?

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,293 @@
+  prefix = "Prefix"
+  ## POST HTTP(or HTTPS) ##
+  # Url name of the Warp 10 server
+  warp_url = "WarpUrl"
```
Let's set the default to `"http://127.0.0.1:8080"`?

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,293 @@
+// Write metrics to Warp10
+func (w *Warp10) Write(metrics []telegraf.Metric) error {
+
+	if len(metrics) == 0 {
```
Instead of this, check if the payload is empty a few lines down and return then.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,293 @@
+	}
+
+	if strings.Contains(body, "Invalid token") {
+		return fmt.Sprintf("Invalid token: %v", w.Token)
```
Don't include the token in any errors because then it will be logged, which we don't want to do with any secrets.

danielnelson commented on plugins/outputs/warp10/warp10.go:
```diff
@@ -0,0 +1,293 @@
+		if len(parts) > 1 {
+			limit = parts[1]
+		}
+		return fmt.Sprintf("MADS exceeded: %v", limit)
```
This will look strange if the regex doesn't match, it may not be worth the complexity to retrieve this information.  If you want to keep the regex though we should compile the regex at package level.  If we switch to using the status line for the error message though I think we can just return the full error.

aurrelhebert commented:
Hello sorry about the delay. 
I just updated this PR to use the HTTP status (or the full body only if requested by the User - which can be useful to debug in case the error differs from the status).

aurrelhebert commented:
Any update?

danielnelson commented:
@aurrelhebert I made a couple tweaks for items I noticed post merge, could you take a look at #6885?  The main changes are limiting uint64 fields to the largest int64 and to use the metric collection time as the sent timestamp, other than that it is just documentation updates.
